File: cpp/132.Palindrome_Partitioning_II.cpp
class Solution {
public:
    int minCut(string s) {
  		if(s.size() <= 1)	return 0;

		// 判断区间[i, j]的字符是否是回文串
  		bool isPal[s.size()][s.size()];

  		// 记录区间[i, size]需要切割的次数
		int cut[s.size()+1];
        cut[s.size()] = -1;
		for(int i = s.size()-1; i >= 0; --i) {
			cut[i] = INT_MAX;
			for(int j = i; j < s.size(); ++j) {
				isPal[i][j] = false;
                if(i == j || i+1 == j && s[i] == s[j] || i+1 < j && isPal[i+1][j-1] && s[i] == s[j]) {
                    isPal[i][j] = true;
					cut[i] = min(cut[i], cut[j+1] + 1);
                }
			}
		}
		return cut[0];
    }
};

File: go/132.Palindrome_Partitioning_II.go
func minCut(s string) int {
	if len(s) <= 1 {
		return 0
	}

	size := len(s)

	// 记录[i, j]是否为回文
	var isPal [][]bool
	for i := 0; i < size; i++ {
		isPal = append(isPal, make([]bool, size))
	}

	// 记录[i, size]分割次数
	cut := make([]int, size+1)
	cut[size] = -1

	for i := size - 1; i >= 0; i-- {
		cut[i] = size
		for j := i; j < size; j++ {
			if (i == j) || (i+1 == j && s[i] == s[j]) || (i+1 < j && isPal[i+1][j-1] && s[i] == s[j]) {
				isPal[i][j] = true
				if cut[i] > cut[j+1]+1 {
					cut[i] = cut[j+1] + 1
				}
			}
		}
	}
	return cut[0]
}

